guide_drafts/schema-in-depth.md

```diff
@@ -0,0 +1,218 @@
+Schema in Depth
+===============
+
+In this guide we're going to look at what exactly `infer_schema!`, `infer_table_from_schema!`, and `table!` do. For `table!` we will show a simplified version of the actual code that gets generated, and explain how each piece is relevant to you. If you've ever been confused about what exactly is getting generated, or what `use schema::posts::dsl::*` means, this is the right place to be.
+
+`infer_schema!` is a macro provided by `diesel_codegen` when you have enabled the feature for one or more database backends. The macro will establish a database connection at compile time, query for a list of all the tables, and generate `infer_table_from_schema!` for each one. `infer_schema` will skip any table names which start with `__`.
+
+`infer_table_from_schema!` has a similar role. It will establish a database connection at compile time, and query the database for the list of columns for the given table. It will then use this information to generate an invocation of `table!` for you.
+
+You can get the same `table!` invocations that `infer_schema!` would have generated by running `diesel print-schema`. If you do not want to require a database with the proper schema be present to compile your app, `diesel print-schema` is the recommended way to create your schema file. Typically these invocations would be placed in `src/schema.rs`.
+
+`table!` is where the bulk of the code gets generated. If you wanted to, you could see the actual exact code that gets generated by running `cargo rustc -- -Z unstable-options --pretty=expanded`. However, the output will be quite noisy, and there's a lot of code that won't actually be relevant to you. Instead we're going to go step by step through a *simplified* version of this output, which only has the code which you would use directly.
```

Comment:
maybe also link to https://github.com/dtolnay/cargo-expand (which `rustfmt`s the `--pretty=expanded` output and even supports colorized output)